### PR TITLE
feat: Show an offline message using a hook to report if app is offline or not

### PR DESCRIPTION
**Pull request by knipec**
## Description
- `isOffline` exists as state in a context provider, so there is a centralized source of truth.
- Setting up the netinfo event listener happens on app start. Listening for netinfo events will stop when the app is not "active", and will resume when the app becomes "active" again. This logic is also centralized in the context provider. The logic itself is heavily based on [Whitespectre's example code here](https://docs.google.com/document/d/1vc2GrmcZLsLPuqeQ5kWJQCxv_8ErfJr6FUR2b3Wr16w/edit)
- There is a trivial hook to use the isOffline state. We use it in the Soil ID sheet to show a "You are offline" message instead of the list of top soil matches.
- Add a small unit test

### Related Issues
Implements #2031

### CODE REVIEW

File: dev-client/src/hooks/connectivityHooks.ts
Comment (knipec) [inline dev-client/src/hooks/connectivityHooks.ts:25]: This feels clunky. Is there a way to ensure the context is hooked up to the state?

File: dev-client/src/context/ConnectivityContext.tsx
Comment (paulschreiber) [inline dev-client/src/context/ConnectivityContext.tsx:75]: issue: remove the debug logging.

File: dev-client/src/context/ConnectivityContext.tsx
Comment (paulschreiber) [inline dev-client/src/context/ConnectivityContext.tsx:42]:
```suggestion
```

File: dev-client/src/screens/LocationScreens/components/soilId/SoilIdMatchesSection.tsx
Comment (tm-ruxandra) [inline dev-client/src/screens/LocationScreens/components/soilId/SoilIdMatchesSection.tsx:171]: I think it would be good to make this a dedicated component in another file.

File: dev-client/src/context/ConnectivityContext.tsx
Comment (tm-ruxandra) [inline dev-client/src/context/ConnectivityContext.tsx:28]: IMO we shouldn't be exposing the state setter via the context like this; if possible it'd make more sense to be read-only, I think. (Unless there's a specific situation you've envisioned for setting it outside of this context provider?)

**Review by tm-ruxandra — CHANGES_REQUESTED** [review]
The overall structure looks good and intuitive! A few comments before I'd want it merged in.

File: dev-client/src/context/ConnectivityContext.tsx
Comment (knipec) [reply dev-client/src/context/ConnectivityContext.tsx:28]: Ah, good call, thanks!

File: dev-client/src/context/connectivity/ConnectivityContext.tsx
Comment (knipec) [inline dev-client/src/context/connectivity/ConnectivityContext.tsx:75]: This line not in the example code from Whitespectre folks. I'm guessing that's just because it's unnecessary if this provider wraps the app, but not harmful, is that correct?

File: dev-client/src/hooks/connectivityHooks.ts
Comment (knipec) [inline dev-client/src/hooks/connectivityHooks.ts:29]: The null check feels a bit clunky. Is there a better way to enforce that we expect the `isOffline` state is created (by the `ConnectivityContextProvider`) before anything else uses the `ConnectivityContext`?

File: dev-client/src/screens/LocationScreens/components/soilId/MessageBoxesInsteadOfSoilMatches.tsx
Comment (paulschreiber) [inline dev-client/src/screens/LocationScreens/components/soilId/MessageBoxesInsteadOfSoilMatches.tsx:25]: suggestion: each of these components should be in its own file.

File: dev-client/src/context/connectivity/connectivityLogic.ts
Comment (paulschreiber) [inline dev-client/src/context/connectivity/connectivityLogic.ts:18]: question: do we want as isOnline function?

Comment (knipec) [reply dev-client/src/context/connectivity/connectivityLogic.ts:28]: Sure, I'll extract it but not export it since nothing outside the file uses it so far

File: dev-client/src/hooks/connectivityHooks.ts
Comment (tm-ruxandra) [reply dev-client/src/hooks/connectivityHooks.ts:29]: Instead of `createContext<ConnectivityContextType | null>(null);` with a nullable context value, you could just set the type to  `SiteUserRole` and have the initial value be some default `ConnectivityContextType` object. (Basically the null-object pattern rather than actual `null`). Not sure what the implications would be of assuming that the context should be online / offline in its default uninitialized state, but presumably it'd be easy to settle on a sensible convention.

File: dev-client/src/context/connectivity/ConnectivityContext.tsx
Comment (tm-ruxandra) [reply dev-client/src/context/connectivity/ConnectivityContext.tsx:76]: I think that's right? But if you're unsure we can wait for comment from them.

**Review by tm-ruxandra — APPROVED** [review]
LGTM (pending any other outside input you want!)

File: dev-client/src/context/connectivity/ConnectivityContext.tsx
Comment (lucaslt89) [reply dev-client/src/context/connectivity/ConnectivityContext.tsx:76]: @knipec that's correct, it's not harmful to have this line and actually a safe net in case you want to unmount the context provider for any reason.

File: dev-client/src/context/connectivity/connectivityLogic.ts
Comment (lucaslt89) [inline dev-client/src/context/connectivity/connectivityLogic.ts:18]: this `getIsOnLine` seems unnecessary? unless you plan to use it in your `ConnectivityContext` later on it could be part of the `getIsOffline` function:
```
export const getIsOffline = ({
  isConnected,
  isInternetReachable,
}: {
  isConnected: boolean | null;
  isInternetReachable: boolean | null;
}) => {
  return !isConnected || !isInternetReachable;
};
```

Comment (knipec) [reply dev-client/src/context/connectivity/connectivityLogic.ts:18]: I mostly just had it for enforcement around when `getIsOffline` should return `null`, and that code has slightly different behavior when `isConnected` is true but `isInternetReachable` is null. I'm not sure how often we'll be in that state or how important it is to track, but since netinfo distinguishes between true / false / null, I figured we should too.

**Pull request merged**
